File: src/services/imageExtractor.ts
import * as cheerio from 'cheerio';
import { fetchUrl } from './httpFetcher';

export interface ExtractedImage {
  src: string;
  alt: string;
  width?: string;
  height?: string;
  loading?: 'lazy' | 'eager' | string;
  dataSrc?: string;
  srcset?: string;
  fileSize?: number; // Size in bytes if available
  mimeType?: string;
  isMissingAlt: boolean;
  isMissingDimensions: boolean;
  isLazy: boolean;
}

export interface ImageExtractionResult {
  totalImages: number;
  missingAltCount: number;
  missingDimensionsCount: number;
  lazyLoadedCount: number;
  images: ExtractedImage[];
}

export function extractImagesFromHtml(html: string, baseUrl?: string): ImageExtractionResult {
  const $ = cheerio.load(html || '');
  const images: ExtractedImage[] = [];

  let missingAltCount = 0;
  let missingDimensionsCount = 0;
  let lazyLoadedCount = 0;

  $('img').each((_, el) => {
    let src = $(el).attr('src')?.trim() || $(el).attr('data-src')?.trim() || '';
    const dataSrc = $(el).attr('data-src')?.trim();
    const srcset = $(el).attr('srcset')?.trim();

    if (src && baseUrl) {
      try {
        src = new URL(src, baseUrl).href;
      } catch {
        // Keep raw src
      }
    }

    const alt = $(el).attr('alt')?.trim() || '';
    const width = $(el).attr('width')?.trim();
    const height = $(el).attr('height')?.trim();
    const loading = $(el).attr('loading')?.trim().toLowerCase();

    const isMissingAlt = alt.length === 0;
    const isMissingDimensions = !width || !height;
    const isLazy = loading === 'lazy' || !!dataSrc || $(el).hasClass('lazyload') || $(el).hasClass('lazy');

    if (isMissingAlt) missingAltCount++;
    if (isMissingDimensions) missingDimensionsCount++;
    if (isLazy) lazyLoadedCount++;

    images.push({
      src,
      alt,
      width,
      height,
      loading,
      dataSrc,
      srcset,
      isMissingAlt,
      isMissingDimensions,
      isLazy
    });
  });

  return {
    totalImages: images.length,
    missingAltCount,
    missingDimensionsCount,
    lazyLoadedCount,
    images
  };
}

/**
 * Optionally checks HEAD/GET headers for image URLs to retrieve file sizes in bytes.
 */
export async function enrichImageFileSizes(images: ExtractedImage[], timeoutMs: number = 5000): Promise<ExtractedImage[]> {
  const enriched = [...images];

  await Promise.all(
    enriched.map(async (img) => {
      if (!img.src || !img.src.startsWith('http')) return;
      try {
        const fetchRes = await fetchUrl(img.src, timeoutMs);
        if (fetchRes.status_code === 200) {
          const contentLength = fetchRes.headers['content-length'];
          if (contentLength) {
            const parsedSize = parseInt(contentLength, 10);
            if (!isNaN(parsedSize)) {
              img.fileSize = parsedSize;
            }
          }
          const contentType = fetchRes.headers['content-type'] || fetchRes.content_type;
          if (contentType) {
            img.mimeType = contentType;
          }
        }
      } catch {
        // Ignore network errors during header size enrichment
      }
    })
  );

  return enriched;
}

File: backend/services/image_extractor.py
import re
import urllib.parse
from typing import Dict, Any, List, Optional
from bs4 import BeautifulSoup

class ImageExtractor:
    """
    Extracts image attributes from HTML:
    - src
    - alt
    - width & height
    - lazy loading (loading="lazy", data-src, lazy classes)
    - srcset
    - file size (if available)
    """

    @staticmethod
    def extract_images_from_html(html: str, base_url: Optional[str] = None) -> Dict[str, Any]:
        soup = BeautifulSoup(html or "", "html.parser")
        images: List[Dict[str, Any]] = []

        missing_alt_count = 0
        missing_dim_count = 0
        lazy_count = 0

        for img in soup.find_all("img"):
            src = (img.get("src") or img.get("data-src") or "").strip()
            data_src = (img.get("data-src") or "").strip() or None
            srcset = (img.get("srcset") or "").strip() or None

            if src and base_url:
                try:
                    src = urllib.parse.urljoin(base_url, src)
                except Exception:
                    pass

            alt = (img.get("alt") or "").strip()
            width = img.get("width")
            height = img.get("height")
            loading = (img.get("loading") or "").strip().lower() or None

            classes = img.get("class") or []
            if isinstance(classes, str):
                classes = classes.split()

            is_missing_alt = len(alt) == 0
            is_missing_dim = not width or not height
            is_lazy = (
                loading == "lazy" or
                data_src is not None or
                "lazyload" in classes or
                "lazy" in classes
            )

            if is_missing_alt:
                missing_alt_count += 1
            if is_missing_dim:
                missing_dim_count += 1
            if is_lazy:
                lazy_count += 1

            images.append({
                "src": src,
                "alt": alt,
                "width": str(width) if width else None,
                "height": str(height) if height else None,
                "loading": loading,
                "dataSrc": data_src,
                "srcset": srcset,
                "isMissingAlt": is_missing_alt,
                "isMissingDimensions": is_missing_dim,
                "isLazy": is_lazy,
                "fileSize": None
            })

        return {
            "totalImages": len(images),
            "missingAltCount": missing_alt_count,
            "missingDimensionsCount": missing_dim_count,
            "lazyLoadedCount": lazy_count,
            "images": images
        }
